File: pk2cmd-gui/src/Operations.java
import javax.swing.*;
import java.awt.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Operations extends SwingWorker<Integer, Integer> {

    private static final int VALUE_ERROR = 111;

    public Integer doInBackground() throws Exception {
        MainWindow.progressBar.setIndeterminate(true);
        MainWindow.container.setCursor(Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR));
        ProcessBuilder pb = new ProcessBuilder(MainWindow.command);
        pb.redirectErrorStream(true);
        Process p = pb.start();
        InputStreamReader isr = new InputStreamReader(p.getInputStream());
        BufferedReader br = new BufferedReader(isr);

        String lineRead;
        while ((lineRead = br.readLine()) != null) {
            System.out.println("Debug: " + lineRead);
            String progressString = lineRead.replaceAll("[^0-9]+", "");

            int progressValue;
            try {
                progressValue = Integer.parseInt(progressString);
            } catch (Exception e) {
                progressValue = VALUE_ERROR;
                MainWindow.logArea.append(lineRead + "\n");
                MainWindow.logArea.setCaretPosition(MainWindow.logArea.getDocument().getLength());
            }

            if (progressValue >= 0 && progressValue <= 100) {
                MainWindow.progressBar.setIndeterminate(false);
                MainWindow.progressBar.setValue(progressValue);
            } else if (progressValue != VALUE_ERROR) {
                MainWindow.logArea.append(lineRead + "\n");
                MainWindow.logArea.setCaretPosition(MainWindow.logArea.getDocument().getLength());
            }

            if (progressValue == 2) {
                if (MainWindow.buttonPressed == MainWindow.buttonTest) {
                    MainWindow.progressBar.setValue(0);
                } else {
                    MainWindow.progressBar.setValue(0);
                    MainWindow.logArea.append(lineRead + "\n");
                    MainWindow.logArea.setCaretPosition(MainWindow.logArea.getDocument().getLength());
                }
            }
        }

        MainWindow.progressBar.setIndeterminate(false);
        MainWindow.container.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
        MainWindow.buttonPressed.setEnabled(true);
        return null;
    }
}

File: src/Operations.kt
import java.awt.Cursor
import java.io.BufferedReader
import java.io.InputStreamReader
import javax.swing.SwingWorker

class Operations : SwingWorker<Int?, Int?>() {

    companion object {
        private const val VALUE_ERROR = 111
    }

    override fun doInBackground(): Int? {
        MainWindow.progressBar.isIndeterminate = true
        MainWindow.container.cursor = Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR)

        val processBuilder = ProcessBuilder(*MainWindow.command)
        processBuilder.redirectErrorStream(true)

        try {
            val process = processBuilder.start()
            val reader = BufferedReader(InputStreamReader(process.inputStream))

            reader.use { br ->
                var line: String?
                while (br.readLine().also { line = it } != null) {
                    println("Debug:: $line")
                    val progressValue = extractProgress(line!!)

                    if (progressValue in 0..100) {
                        MainWindow.progressBar.isIndeterminate = false
                        MainWindow.progressBar.value = progressValue
                    } else if (progressValue != VALUE_ERROR) {
                        appendToLog(line)
                    }

                    if (progressValue == 2) {
                        if (MainWindow.buttonPressed === MainWindow.buttonTest) {
                            MainWindow.progressBar.value = 0
                        } else {
                            resetProgressBar(line)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            handleError(e)
        } finally {
            MainWindow.progressBar.isIndeterminate = false
            MainWindow.container.cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
            MainWindow.buttonPressed?.isEnabled = true
        }

        return null
    }

    private fun extractProgress(line: String): Int {
        if (line.contains("No PICkit", ignoreCase = true)) {
            resetProgressBar(line)
            return VALUE_ERROR
        }

        val progressString = line.replace("[^0-9]+".toRegex(), "")
        return try {
            progressString.toInt()
        } catch (_: Exception) {
            MainWindow.logArea.append(line + "\n")
            MainWindow.logArea.caretPosition = MainWindow.logArea.document.length
            VALUE_ERROR
        }
    }

    private fun appendToLog(line: String) {
        MainWindow.logArea.append(line + "\n")
        MainWindow.logArea.caretPosition = MainWindow.logArea.document.length
    }

    private fun resetProgressBar(line: String) {
        MainWindow.progressBar.value = 0
        MainWindow.logArea.append(line + "\n")
        MainWindow.logArea.caretPosition = MainWindow.logArea.document.length
    }

    private fun handleError(e: Exception) {
        MainWindow.logArea.append("Error: ${e.message}\n")
        MainWindow.logArea.caretPosition = MainWindow.logArea.document.length
    }
}
